python/ray/workflow/workflow_executor.py

```diff
@@ -0,0 +1,357 @@
+import time
+from collections import defaultdict
+
+import ray
+from ray.exceptions import RayTaskError, RayError
+
+from ray.workflow.common import (
+    WorkflowRef,
+    WorkflowExecutionMetadata,
+    WorkflowStatus,
+    StepID,
+)
+from ray.workflow.exceptions import WorkflowCancellationError, WorkflowExecutionError
+from ray.workflow.step_executor import get_step_executor, _BakedWorkflowInputs
+from ray.workflow.workflow_state import WorkflowExecutionState, TaskExecutionMetadata
+
+if TYPE_CHECKING:
+    from ray.workflow.workflow_context import WorkflowStepContext
+    from ray.workflow.workflow_storage import WorkflowStorage
+
+logger = logging.getLogger(__name__)
+
+
+class WorkflowExecutor:
+    def __init__(
+        self,
+        state: WorkflowExecutionState,
+    ):
+        """The core logic of executing a workflow.
+
+        This class is responsible for:
+
+        - Dependency resolving.
+        - Task scheduling.
+        - Reference counting.
+        - Garbage collection.
+        - Continuation handling and scheduling.
+        - Error handling.
+        - Responding callbacks.
+
+        It borrows some design of event loop in asyncio,
+        e.g., 'run_until_complete'.
+
+        Args:
+            state: The initial state of the workflow.
+        """
+        self._state = state
+        self._completion_queue = asyncio.Queue()
+        self._task_done_callbacks: Dict[StepID, List[asyncio.Future]] = defaultdict(
+            list
+        )
+
+    def is_running(self) -> bool:
+        """The state is running, if there are tasks to be run or running tasks."""
+        return bool(self._state.frontier_to_run or self._state.running_frontier)
+
+    def get_state(self) -> WorkflowExecutionState:
+        return self._state
+
+    @property
+    def output_task_id(self) -> StepID:
+        return self._state.output_task_id
+
+    async def run_until_complete(
+        self, job_id: str, context: "WorkflowStepContext", wf_store: "WorkflowStorage"
+    ):
+        """Drive the state util it completes.
+
+        Args:
+            job_id: The Ray JobID for logging properly.
+            context: The context of workflow execution.
+            wf_store: The store for the workflow.
+
+        # TODO(suquark): move job_id inside context
+        """
+        workflow_id = context.workflow_id
+        wf_store.update_workflow_status(WorkflowStatus.RUNNING)
+        logger.info(f"Workflow job [id={workflow_id}] started.")
+
+        self._state.construct_scheduling_plan(self._state.output_task_id)
+        self._state.init_context(context)
+
+        while self.is_running():
+            # ------------ poll queued tasks ------------
+            queued_tasks = self._poll_queued_tasks()
+
+            # --------------- submit task ---------------
+            for task_id in queued_tasks:
+                self._submit_task(task_id, job_id=job_id)
+                self._post_process_submit_task(task_id, wf_store)
```

Comment:
Why not just put post_process_submit_task into submit_task?

Reply:
I mean why do we need two functions here? When I'm reading post_process_submit_task, I'm not sure what do we do just by the name.
Another option is to make the name more explicit to reflect what's the thing going on there.